### PR TITLE
State sync with TestLoop with memtries.

### PR DESCRIPTION
**Pull request by robin-near**
This PR makes it possible to use TestLoop to test state sync behavior, including memtrie compatibility with single-shard tracking. It is very big, but consists of the following changes:

* SyncAdapter now takes a function for how to create a SyncActor; for production, it spawns an Arbiter, starts the actor in it, and returns a handle for how to send messages to the actor and how to shut down the arbiter. For TestLoop, it creates the actor, and adds it to the hashmap of shard -> SyncActor in the testloop data. This way, for testing we don't spawn additional threads.
* State sync dumper undergoes a similar change. We pass in a function that is responsible for creating the thread used for dumping (takes a future and returns a cleanup function). For production, it spawns an arbiter, posts the future in the arbiter, and returns a function to stop the arbiter; for testloop, we just spawn the future into the event loop. Also, state sync dumper is converted to an object instead of a function so that it's easier to manage in the TestLoop setup (just call start()).
* StateSnapshotActor undergoes the standard actor change, where we use multi-senders instead of actix::Addr.

There are some issues that make this test slow:

* State snapshot creation does not work with `create_test_store()`, because it relies on rocksdb features. Without knowing how it works, I've opted to just run the test with real RocksDB in a temp directory. The drawback of this is that it makes the test slow. It is so slow due to RocksDB and state snapshot creation that the test now takes 37s. Ideally, we would want to make `create_test_store()` work with snapshots.
* Another issue with test is what has been brought up before: validator mandate threshold is too small and it makes the test slow. This hopefully will be addressed later when validator selection changes.
* I can't really understand the state sync timeout logic. When the state sync code tries to read a header from local disk and it doesn't exist, instead of failing and triggering a retry, it doesn't seem to do anything and we have to rely on the timeout logic to trigger another retry. Something is a bit off here - I don't see that the external storage code path sets "download.error = true" and even when I set that it doesn't seem to be true. For now I've set the timeout to 100ms for the test loop test, but this perhaps is worth looking into.

### CODE REVIEW

**Comment by robin-near** [comment]
@pugachAG @Longarithm This is now ready for review.

File: chain/client/src/sync/adapter.rs [hunk 1]
```diff
@@ -31,86 +29,108 @@ pub enum SyncMessage {
...
     /// Address of the sync actors indexed by the SharUid
-    actor_handler_map: HashMap<ShardUId, ActorHandler>,
+    actor_handler_map: HashMap<ShardUId, SyncActorHandler>,
     /// Message channel for client
     client_adapter: Sender<SyncMessage>,
     /// Message channel with network
     network_adapter: Sender<PeerManagerMessageRequest>,
+    /// Function to make an actor
+    actor_maker: Arc<
+        dyn Fn(ShardUId, Sender<SyncMessage>, Sender<PeerManagerMessageRequest>) -> SyncActorHandler
```
Comment (pugachAG) [inline chain/client/src/sync/adapter.rs:55]: nit: `Fn(ShardUId, Sender<SyncMessage>, Sender<PeerManagerMessageRequest>) -> SyncActorHandler  + Send + Sync` is used in multiple places, so maybe let's use type alias for that

**Review by pugachAG — APPROVED** [review]
LGTM! probably a good idea to also get a review from someone who worked on state sync: cc @VanBarbascu

File: core/primitives/src/challenge.rs [hunk 2]
```diff
@@ -23,6 +24,18 @@ impl Default for PartialState {
     }
 }
 
+// When debug-printing, don't dump the entire partial state; that is very unlikely to be useful,
+// and wastes a lot of screen space.
+impl Debug for PartialState {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            PartialState::TrieValues(values) => {
```
Comment (pugachAG) [inline core/primitives/src/challenge.rs:32]: nit: I believe `write!(f, {} trie values", values.len())` should work here

**Pull request merged**